### ISSUE TITLE
Fixed performance on the Modules listing page /admin/modules

### ISSUE THREAD

**Issue by alanmels** [backdrop/backdrop-issues#5046]
**Description of the bug**

I've been always bothered by how quickly the Functionality (admin/modules) page gets really slow with larger number of contributed modules enabled. So I decided to find out the culprit. 

**Steps To Reproduce**

1. Enable the Devel module;
2. Open the `system.module`, find the `system_rebuild_module_data()` function and put `sdpm($modules_cache);` like so:

```
function system_rebuild_module_data() {
  $modules_cache = &backdrop_static(__FUNCTION__);
  sdpm($modules_cache);
```
Refresh the modules page and see `$modules_cache` always returns empty, so the if condition further below is completely useless. 

However, caching for the modules page starts working as expected if the `$modules_cache = &backdrop_static('__FUNCTION__');` is replaced with `$modules_cache = &backdrop_static('modules_cache');` like so:

```
function system_rebuild_module_data() {
  $modules_cache = &backdrop_static('modules_cache');
  // Only rebuild once per request. $modules and $modules_cache cannot be
  // combined into one variable, because the $modules_cache variable is reset by
  // reference from system_list_reset() during the rebuild.
  if (!isset($modules_cache)) {
    $modules = _system_rebuild_module_data();
    ksort($modules);
    system_get_files_database($modules, 'module');
    system_update_files_database($modules, 'module');
    $modules = _module_build_dependencies($modules);
    $modules_cache = $modules;
  }
  return $modules_cache;
}
```

Finally, I have really fast Modules page.

To describe the problem: Two modules try to reset the caches on /admin/modules.

- The system module (which provides the page)
- The update module (which has nothing to do with this page)

To my understanding this happened by mistake (many years back) and should get fixed, so that only the module that's _responsible_ for that page flushes caches on that page.

---
PR by @docwilmot https://github.com/backdrop/backdrop/pull/3601

**Comment by indigoxela** [comment]
Interesting... I could verify this with debug():

```
function system_rebuild_module_data() {
  $modules_cache = &backdrop_static(__FUNCTION__);
  debug('Function gets called: ' . __FUNCTION__);
  // Only rebuild once per request. $modules and $modules_cache cannot be
  // combined into one variable, because the $modules_cache variable is reset by
  // reference from system_list_reset() during the rebuild.
  if (!isset($modules_cache)) {
    $modules = _system_rebuild_module_data();
    ksort($modules);
    system_get_files_database($modules, 'module');
    system_update_files_database($modules, 'module');
    $modules = _module_build_dependencies($modules);
    $modules_cache = $modules;
    debug('Code runs, no cache.');
  }
  return $modules_cache;
}
```
Although __FUNCTION__ is the same, subsequent calls to system_rebuild_module_data() still run the code inside the if clause, so there's effectively no caching.

Changing `__FUNCTION__` to the string `system_rebuild_module_data` (note, still the function name), and the cache still isn't working as expected.
Changing the string to `foobarbazbla` and the cache works fine. :confused:

For what it's worth, if I change it to `&backdrop_static(__FUNCTION__ . 'bla')` caching also works.

What's so problematic with "system_rebuild_module_data", that it breaks the static cache? Can someone with better knowledge of the backdrop_static cache shed a light on this, please?

**Comment by indigoxela** [comment]
Ha! There it is: https://github.com/backdrop/backdrop/blob/a259d2d276e5f43db925a408a2f06944eee25324/core/includes/module.inc#L230

The cache is ineffective, because it gets reset.

**Comment by klonos** [comment]
@alanmels I would look for calls to `backdrop_static_reset('system_rebuild_module_data');` or `system_list_reset()` to blame.

**Comment by klonos** [comment]
@indigoxela jinx! 😆

**Comment by klonos** [comment]
...I was looking through d.org issues like the following:

- https://www.drupal.org/project/drupal/issues/794936
- https://www.drupal.org/project/drupal/issues/651086
- https://www.drupal.org/project/drupal/issues/1436684#comment-6575322

**Comment by klonos** [comment]
...it seems quite likely to me that we may be too overzealous at some place, and we might be doing `system_list_reset()` instead of just reseting say only the theme cache. But that's just a though/hunch.

**Comment by docwilmot** [comment]
This is an interesting conundrum here. I have no clue how to resolve this.

FYI since it hasnt been explicitly stated, `system_list_reset()` is being called at the end of `system_update_files_database()` which is itself called from within `system_rebuild_module_data()`, thus immediately resetting the static cache each time.

**Comment by alanmels** [comment]
Thank you for the inputs. It seems there is more here than just replacing a single string. I'm not sure anymore what would be the ideal fix, but let's get this - seemingly long time D7 problem - finally fixed.

**Comment by alanmels** [comment]
> Ha! There it is: https://github.com/backdrop/backdrop/blob/a259d2d276e5f43db925a408a2f06944eee25324/core/includes/module.inc#L230
> 
> The cache is ineffective, because it gets reset.

I've reverted the change I made, then commented out the `backdrop_static_reset('system_rebuild_module_data');` line in the `system_list_reset()` function and, unfortunately, the problem is still there - the modules page is called twice, because caching is not working.

**Comment by klonos** [comment]
>This is an interesting conundrum here. I have no clue how to resolve this.

Same here. Seems like a catch 22.

**Comment by klonos** [comment]
...I just compared `system_list_reset()`, `system_update_files_database()`, and `system_rebuild_module_data()` between D7 and Backdrop, and they are basically identical. So if we are having this issue, then D7 has the same issue. Perhaps it's worth opening an issue in d.org with our findings, and allow the collective minds of our brethren to chime in with thoughts/ideas on how this could be solved 🤔

**Comment by alanmels** [comment]
@klonos, that would be nice if it really helps. However, we all know some issues on D take forever to be properly resolved. It's basically the same issue that you've referenced earlier https://www.drupal.org/project/drupal/issues/794936, however it was for D6, never resolved and now is closed.

Another pointer that could help: the `admin/modules/list` page is always faster, because it loads modules only once. So if fixing the issue for `admin/modules` is too cumbersome, then we could point all the menu links that previously lead to `admin/modules` to `admin/modules/list` instead.

**Comment by alanmels** [comment]
Interestingly, the `admin/appearance/list` page also always reads from cache, however `admin/appearance` always goes through the broken `system_rebuild_module_data()` function. 

So then the naming is little controversial - if the function servers not only modules, but themes page then maybe something like `system_rebuild_project_data()` would be a better fit?

**Comment by indigoxela** [comment]
@klonos Waiting for Drupal to resolve something means waiting for a long time, potentially forever. I wouldn't rely on that.

@alanmels Whatever we figure out as a solution, two things are important:

1. We must not rename functions that are potentially used by contrib modules (it would break them)
2. We must not break cache flushing, because for a real rebuild (module enable/disable) this is important

And now I'm curious, who can solve our nice little catch 22. :smirk:

**Comment by klonos** [comment]
>that would be nice if it really helps. However, we all know some issues on D take forever to be properly resolved.

> Waiting for Drupal to resolve something means waiting for a long time, potentially forever. I wouldn't rely on that.

I never implied to _wait_ for this to be fixed in D7. Just to start a discussion there, and see if more brainpower can help come up with a solution.

> So then the naming is little controversial...

> We must not rename functions that are potentially used by contrib modules (it would break them)

I agree with both these points ^^ ...and we have a place to discuss trying to bring consistency to these functions: #2548

**Comment by indigoxela** [comment]
admin/modules - unlike the .../list page this page runs system_rebuild_module_data twice, the tracebacks are:

```
array (
  0 => 'system_rebuild_module_data',
  1 => 'update_get_projects',
  2 => 'update_get_available',
  3 => 'update_requirements',
  4 => 'update_init',
  5 => 'module_invoke_all',
  6 => '_backdrop_bootstrap_full',
  7 => 'backdrop_bootstrap',
)

array (
  0 => 'system_rebuild_module_data',
  1 => 'system_modules',
  2 => 'backdrop_retrieve_form',
  3 => 'backdrop_build_form',
  4 => 'backdrop_get_form',
  5 => 'system_block_view',
  6 => 'module_invoke',
  7 => 'buildBlock',
  8 => 'getContent',
  9 => 'renderBlock',
  10 => 'renderBlocks',
  11 => 'renderLayout',
  12 => 'render',
  13 => 'layout_route_handler',
  14 => 'menu_execute_active_handler',
)
```

^^ The first of the above calls does happen on admin/modules/list.

**Comment by alanmels** [comment]
So since `admin/modules/list` is faster, could we replace the menu link to it while it's not clear how long resolving the problem for `admin/modules` would take?

**Comment by alanmels** [comment]
@indigoxela, while I agree that, if ever, the change in the function's name must be done with caution:

>     1. We must not rename functions that are potentially used by contrib modules (it would break them)

the `system_rebuild_module_data()` is a core function that is needed for the specific feature, so I doubt there are large number of contributed modules which ever interacted with it. If few modules indeed did, then it's easy for them to adjust their code.

Generally and not only for this specific case, I wish Backdrop community were bolder to make changes in core that make sense sooner then leaving them hanging around for years as it often happened for Drupal 6/7. 
 
>     2. We must not break cache flushing, because for a real rebuild (module enable/disable) this is important

Isn't it broken already? As I noted in the very beginning the `$modules_cache` int the `system_rebuild_module_data()` function always returns empty, so the `if` condition is completely useless. Also, I've made sure to enable/disable modules after the suggested change was made and the modules list properly gets rebuilt.

**Comment by docwilmot** [comment]
I'm a bit surprised that `admin/modules/list` does a different thing from `admin/modules`. Both paths load the same page callback which is `system_modules()`. How are we testing this?

**Comment by alanmels** [comment]
> I'm a bit surprised that `admin/modules/list` does a different thing from `admin/modules`. Both paths load the same page callback which is `system_modules()`. How are we testing this?

Just place `sdpm('test');` in the function like so:

```
function system_rebuild_module_data() {
  sdpm('test');
```
and you will see duplicated output on `admin/modules` whereas the `admin/modules/list` always returns single output.

**Comment by indigoxela** [comment]
> I'm a bit surprised that admin/modules/list does a different thing from admin/modules. Both paths load the same page callback which is system_modules(). How are we testing this?

@docwilmot look at my [previous comment](https://github.com/backdrop/backdrop-issues/issues/5046#issuecomment-821960413). I did a debug_backtrace() in function system_rebuild_module_data() and extracted the functions that led there to display the trace with debug().

I'm not sure if there's a way to fix that, which isn't too radical. Didn't dig deeper yet, though.

**Comment by klonos** [comment]
> Generally and not only for this specific case, I wish Backdrop community were bolder to make changes in core that make sense sooner then leaving them hanging around for years as it often happened for Drupal 6/7.

@alanmels our first principle is backwards compatibility: https://backdropcms.org/philosophy#principles

We cannot remove/rename functions, but what we can do is introduce new ones with the proper names, and then have the old ones "redirect" to the new ones, along with a `watchdog_deprecated_function()`. Either way, not a matter of discussion for this issue feel free to add a comment in #2548 or even start working on it 😉

**Comment by indigoxela** [comment]
Not sure if it helps, but the double call to function system_rebuild_module_data() is caused by [update_init()](https://github.com/backdrop/backdrop/blob/a259d2d276e5f43db925a408a2f06944eee25324/core/modules/update/update.module#L91).

Potentially a small change there could fix the problem, but that needs more recherche.

**Comment by indigoxela** [comment]
I created a PoC for further discussions. My approach is to prevent double runs as described in [my above comment](https://github.com/backdrop/backdrop-issues/issues/5046#issuecomment-821960413).

Two modules, system and update, are flushing data on page "admin/modules", once is enough.

Another finding (also by @alanmels BTW): the static cache within function system_rebuild_module_data() is useless and could get removed.

How to test my alternate [PR](https://github.com/backdrop/backdrop/pull/3600):

Put in a `debug()` in function system_rebuild_module_data() and verify that it does run once - and only once - on pages that should never show stale data, "admin/modules" is one of them.

**Comment by alanmels** [comment]
@indigoxela bellissimo! I've just tested your PR and it does the fixing. Also I've adjusted my own PR per your findings - it simplifies the `system_rebuild_module_data()` function by removing the useless lines.

**Comment by indigoxela** [comment]
@alanmels many thanks for testing this!

I guess, we're creating confusion with our two PRs, which solve different things. We probably better split this into two individual issues. One for the performance task (prevent multiple runs of same function) and one for the futile static cache usage (which might a little more discussion and evaluation).

To me it seems, that this one should be the "performance issue". What do other people think?

**Comment by alanmels** [comment]
Please also read https://backdrop.zulipchat.com/#narrow/stream/218635-Backdrop/topic/zen.2Eci.20is.20unstable

**Comment by alanmels** [comment]
Please go ahead and create another issue. I understand you're trying to move along more systematic ways. I'll be happy the duplicated load issue for Modules page will be finally resolved (after waiting for so many years for Drupal 6 and Drupal7).

**Comment by docwilmot** [comment]
Or even better: in `system_update_files_database()`, dont call `system_list_reset()`, just add the function calls from `system_list_reset()` directly (minus  `backdrop_static_reset('system_rebuild_module_data')` of course):
```
function system_update_files_database(&$files, $type) {
  $result = db_query("SELECT * FROM {system} WHERE type = :type", array(':type' => $type));
   .....

  backdrop_static_reset('system_list');
  backdrop_static_reset('list_themes');
  cache('bootstrap')->deleteMultiple(array('bootstrap_modules', 'system_list'));
}
```

Tested, works.

**Comment by alanmels** [comment]
> @alanmels I also dont think removing the static cache there is best either. Fact is its an expensive function, we agree, and though the `admin/modules` page is _one_ path that we know its called twice, we cant be certain that it wont be called twice in other situations.

I see your concern. But give me one example, when `$modules_cache = &backdrop_static(__FUNCTION__);`line works. I've run many tests, but it always return empty. So what's the point of setting it if it's always empty?

**Comment by docwilmot** [comment]
Did you look at https://github.com/backdrop/backdrop/pull/3601?

**Comment by alanmels** [comment]
> Did you look at [backdrop/backdrop#3601](https://github.com/backdrop/backdrop/pull/3601)?

I've just tested https://github.com/backdrop/backdrop/pull/3601 and, unfortunately, with it the module's list is loaded twice.

**Comment by docwilmot** [comment]
Just to be clear, with this PR the function will be called twice on `admin/modules` as usual but the rebuild will only happen the first call. Try putting a `dpm()` after the line `if (!isset($modules_cache)) {`. I've tested this in a custom module with multiple calls to `system_rebuild_module_data()`; the rebuild only happens once. and the page load is instant.

> Additionally, clearing cache is not working. Reverting makes clearing cache work normally.

I dont get what you mean by this.

**Comment by klonos** [comment]
Of the 3 approaches, @docwilmot's seems the safest to me (although perhaps not as elegant, as it duplicates code from an existing function).

**Comment by indigoxela** [comment]
I'll try to defend my approach: To me it makes no sense to call an expensive function multiple times. We're able to identify, why it's called twice, so to me it makes most sense to remove the redundant call. Once is enough. That has no side effects at all - correct me if I'm wrong. Two modules reset things on the /admin/modules page, but as the system module defines that path in its hook_menu, it seems most appropriate that the system module takes care of providing fresh data - and it does.

Proceeding with calling the function twice on that page just seems so odd to me - we know that it's not necessary and actually does harm performance-wise. So let's stop doing that. IMO this is the only clean solution in this case.

**Comment by alanmels** [comment]
I share @indigoxela's opinion. Call the same function twice on the same page just doesn't make sense. Also I said with @docwilmot something goes awry with clearing cache - it gave WSOD to me.

**Comment by docwilmot** [comment]
We call the same function twice (much more than twice even) all the time in Backdrop. Think how many times `node_load()` or `user_load()` or `entity_get_info()` could get called for the same node or user on a page. Thats not a problem if the result of the function is cached, which is what we do in all those situations. The 2nd (or 3rd or 4th ...) call is irrelevant, as no processing happens. My PR fixes this for that function.

@indigoxela PR is perfectly reasonable, but I have questions about some processing that is left out. If this is not a problem, then I agree we could use that solution.

However, as I said before, we are only now seeing this one situation where this expensive function is being called twice. Some contrib in the future may result in a double call. We cannot avoid that sometimes. Fixing the static cache seems to me the ideal solution.

**Comment by indigoxela** [comment]
> We call the same function twice (much more than twice even) all the time in Backdrop.

That's correct, but then the static cache should work properly - and I don't think that this situation is a valid use-case for static caching anyway.

That system_rebuild_module_data() runs twice on /admin/modules but **not** on /admin/modules/list clearly shows that this happens by mistake.

Can we please create a follow-up issue for the "static cache or not in system_rebuild_module_data" discussion? I already suspected, that there will be objections. And here it would be a blocker for a reasonable and logic performance bug fix.

**Comment by docwilmot** [comment]
> That's correct, but then the static cache should work properly

My PR literally fixes this exact problem. While solving the other issue of performance.

But I'm going to step slowly away from this discussion and hope the core committers can find the best solution.

**Comment by alanmels** [comment]
@docwilmot, right about multiple runs of the same helper functions. But not as yourself put so "expensive" heavy function as clearing cache. So instead of adding the conditions not to trigger certain parts of the same conditions, the logic tells just to avoid running through it twice.

My understanding is that when visiting the `admin/modules` page, the `system_rebuild_module_data()` is triggered and whatever was needed from the function is already satisfied by that point. And there is absolutely no need for another run of the same function. 

As for:

> On certain paths Update module requires certain rebuild functions to run.

we should really operate based not on guesses, but specific examples. I've run number of scenarios to trigger the Update module (such as deliberately downloading the older version of modules bot core and contributed modules) and seen no problem at all.  If you give at least one of those `certain paths`, then I'm happy to run additional tests. But even in that unlikely scenario I would think it's a bug that needs to be fixed elsewhere.

**Comment by klonos** [comment]
I plan to bring this during the next weekly meeting, and ask the (now extended 😄 ...shhhh!) team of core committers to chime in and move this forward.

**Comment by alanmels** [comment]
Any updates on this? It would be really nice to get this fixed.

**Comment by indigoxela** [comment]
Again a hint, why I think the double run happens by mistake:

The system module already takes care of preventing stale data on /admin/modules _and_ /admin/modules/list - which are basically the same page on different paths.

The updater module flushes caches, too, on page /admin/modules, but not on /admin/modules/list.

So, here's a provocative (ironic) suggestion: If people are so convinced that two modules should flush caches /admin/modules, this also must happen on /admin/modules/list, right?

In order to also slow down that page, which is currently faster than /admin/modules - as function system_rebuild_module_data() runs only once there. :snail: 

But my real (serious) suggestion is to run that function only once on /admin/modules - that's what my PR tries to achieve.

**Comment by alanmels** [comment]
@indigoxela, I've noticed you closed https://github.com/backdrop/backdrop/pull/3600#issuecomment-834228756 With all due respect to practices around here, I still scratch my head when some issues are decided by the prevalence of just one opinion. What about Backdrop consensus? Maybe you should've wait for review of more core committers? I wonder what @quicksketch and others think about this issue?

**Comment by indigoxela** [comment]
> What about Backdrop consensus?

There is no consensus and it seems impossible to me. @docwilmot insists on multiple runs with (repaired) static cache, I think the only clean solution is running that cache refresh only once.

I won't block his approach, but I won't do anything to get it in.

I tried to get attention from the core committers (via chat), but they completely missed the point. We're on our own, more precisely, you're on your own, because I'm out here, too. :four_leaf_clover:

**Comment by quicksketch** [comment]
Hi folks, I apologize that I haven't given this the time to review fully. I didn't realize this issue was at an impasse.

I read through this entire issue and the separate discussion on @indigoxela's PR. I left a comment there on that approach: https://github.com/backdrop/backdrop/pull/3600#issuecomment-835573932

The takeaway here is that I think https://github.com/backdrop/backdrop/pull/3600 unintentionally changes the behavior of Update module (I posted more about that [in the comment](https://github.com/backdrop/backdrop/pull/3600#issuecomment-835573932)). I don't see a way to avoid calling `system_rebuild_module_data()` twice and maintain the intended behavior, so I would be inclined to fix the static cache as @docwilmot suggests in https://github.com/backdrop/backdrop/pull/3601. That fix could also be beneficial in the event `system_rebuild_module_data()` is called twice on any other pages in Backdrop (though we don't have any other known examples).

Another alternative is that we _intentionally_ stop regenerating Update caches on various admin pages: Remove the `hook_init()` entirely from Update module and only show messages about updates in relevant places like the Available Updates page and the Status report. That would prevent the expensive rebuilding from happening on the modules page and all the other pages that are listed in `update_init()`. I would be in favor of that, but I think that's a different discussion than this performance-related request.

So to move this request forward, I think https://github.com/backdrop/backdrop/pull/3601 should be merged to solve the performance issue. https://github.com/backdrop/backdrop/pull/3593 should be closed. Looks like @indigoxela already closed https://github.com/backdrop/backdrop-issues/issues/5052 as well.

If it's desired to try avoiding Update module's expensive recalculations on various admin pages, we should open a new ticket for that.

**Comment by quicksketch** [comment]
I filed a new issue to discuss refactoring when we show update notifications at https://github.com/backdrop/backdrop-issues/issues/5083.

**Comment by alanmels** [comment]
@quicksketch I supposed there is a reason behind @docwilmot's insisting on his take of the issue, but really wanted to see some kind of proof-case. Which you've given on https://github.com/backdrop/backdrop/pull/3600#issuecomment-835573932 with plenty of details. Still I find your alternative suggestion to solve this much appealing: displaying update messages on the Available Updates and the Status page would be just enough as users never go to the Modules page to check updates, but other reasons as enabling, disabling, reviewing the modules.

**Comment by klonos** [comment]
Code in https://github.com/backdrop/backdrop/pull/3601 looks good to me 👍🏼

**Comment by klonos** [comment]
...not sure how to best test this? ...do we need to install heaps of modules, and then benchmark how fast the modules pages loads before/after the fix?

**Comment by docwilmot** [comment]
> ...not sure how to best test this 🤷🏼 ...do we need to install heaps of modules, and then benchmark how fast the modules pages loads before/after the fix?

We just need to test that the static cache on `system_rebuild_module_data()` works now with this PR. So a `dpm()` after the line `if (!isset($modules_cache)) {` and a few consecutive calls to`system_rebuild_module_data()` should only show one hit with this PR, whereas without this PR each call would show a hit.

**Comment by stpaultim** [comment]
@klonos - Can an issue be RTBC and "PR-Needs Testing"?

**Comment by klonos** [comment]
@stpaultim yes. The RTBC acronym has 2 meanings:
- in Drupal-land it means "reviewed and tested by the community", which covers both code review and testing
- in Backdrop, since we have a separate "works for me" label specifically for testing, RTBC means (code) "ready to be committed"

My understanding is that one person can review the code, and if no issues are found with coding standards, and it makes sense, they can mark it as RTBC. Another person can then do the manual testing, and add the WFM label.

**Comment by stpaultim** [comment]
@klonos - I see that the logic. However, I find it very confusing. Because, how can something be "Ready to be Committed" if it hasn't been tested yet. 

I've always assumed that the "RTBC" should only be applied after "PR-Works for Me." Until we an issue is marked "Works for Me," we can't really be sure that the code is final yet, let alone RTBC? 

(We should probably discuss this elsewhere, but I'm not sure where?)

**Comment by docwilmot** [comment]
Instructions for quick test right here: https://github.com/backdrop/backdrop-issues/issues/5046#issuecomment-907529850
